src/rust/engine/src/rule_graph.rs

```diff
@@ -837,8 +820,10 @@ fn rhs_for_select(tasks: &Tasks, subject_type: TypeId, select: &Select) -> Entri
   if externs::satisfied_by_type(&select.product, &subject_type) {
     // NB a matching subject is always picked first
     vec![Entry::new_subject_is_product(subject_type)]
+  } else if let Some(&(ref key, _)) = tasks.gen_singleton(&select.product) {
```

Comment:
Yep. My issue is that those noop declarations and their associated dispatch paths represent the same piece of knowledge, but are not tied together.
When we clear up the todo around the dispatch, I think we should make sure to tie them together somehow.